velox/exec/Window.cpp

```diff
@@ -379,25 +425,88 @@ void Window::callApplyForPartitionRows(
         }
         break;
       }
-      case core::WindowNode::BoundType::kPreceding:
-      case core::WindowNode::BoundType::kFollowing:
-        VELOX_NYI("Not supported");
+      case core::WindowNode::BoundType::kPreceding: {
+        if (frameArg->index == kConstantChannel) {
+          auto constantOffset =
+              frameArg->value->template as<ConstantVector<int64_t>>()->valueAt(
+                  0);
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow - constantOffset - firstPartitionRow);
+        } else {
+          readFrameValues(frameArg.value());
+          auto offsetsVector = frameArg->value->asFlatVector<int64_t>();
+          for (auto i = 0; i < numRows; i++) {
+            rawFrameBounds[i] =
+                (startRow + i) - offsetsVector->valueAt(i) - firstPartitionRow;
+          }
+        }
+        break;
+      }
+      case core::WindowNode::BoundType::kFollowing: {
+        if (frameArg->index == kConstantChannel) {
+          auto constantOffset =
+              frameArg->value->template as<ConstantVector<int64_t>>()->valueAt(
+                  0);
+          std::iota(
+              rawFrameBounds,
+              rawFrameBounds + numRows,
+              startRow + constantOffset - firstPartitionRow);
+        } else {
+          readFrameValues(frameArg.value());
+          auto offsetsVector = frameArg->value->asFlatVector<int64_t>();
+          for (auto i = 0; i < numRows; i++) {
+            rawFrameBounds[i] =
+                (startRow + i) + offsetsVector->valueAt(i) - firstPartitionRow;
+          }
+        }
+        break;
+      }
       default:
         VELOX_USER_FAIL("Invalid frame bound type");
     }
   };
 
+  // Updates the start and end frame bounds to a lower bound of 0 and upper
```

Comment:
Nit : Rephrase
"k preceding and k following bounds in ROWS mode can go over the partition limits. So they are bound to the first and last partition rows if required. If the frame is empty incase the frameEnd > frameStart values, then the bounds are set to -1. "